amd_openvx_extensions/amd_nn/src/topk_layer.cpp

```diff
@@ -0,0 +1,301 @@
+    float *x_tensor = new float[count_input_dims_0];
+    //temporary moving ptr
+    float *x_tensor_temp = x_tensor;
+
+    status = vxMapTensorPatch((vx_tensor)parameters[0], num_of_dims, nullptr, nullptr, &map_id, stride, (void **)&ptr_input_0, usage, VX_MEMORY_TYPE_HOST, 0);
+    if(status)
+    {
+        std::cerr << "ERROR: vxMapTensorPatch() failed for input#1 (" << status << ")" << std::endl;
+        return -1;
+    }
+
+    memcpy(x_tensor_temp, ptr_input_0, (count_input_dims_0*sizeof(float)));
+ 
+
+    status = vxUnmapTensorPatch((vx_tensor)parameters[0], map_id);
+
+    if(status) 
+    {
+        std::cerr << "ERROR: vxUnmapTensorPatch() failed for input#5 (" << status << ")" << std::endl;
+        return -1;
+    }
+
+    ERROR_CHECK_STATUS(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_DIMS, input_dims_1, sizeof(input_dims_1)));
+    ERROR_CHECK_STATUS(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_NUMBER_OF_DIMS, &num_of_dims, sizeof(num_of_dims)));
+
+    
+    int64_t * ptr_input_1;
+    vx_size count_input_dims_1 = input_dims_1[0];
+    int64_t *k_tensor = new int64_t[count_input_dims_1];
+
+    status = vxMapTensorPatch((vx_tensor)parameters[1], num_of_dims, nullptr, nullptr, &map_id, stride, (void **)&ptr_input_1, usage, VX_MEMORY_TYPE_HOST, 0);
+    if(status)
+    {
+        std::cerr << "ERROR: vxMapTensorPatch() failed for input#2 (" << status << ")" << std::endl;
+        return -1;
+    }
+
+    memcpy(k_tensor, ptr_input_1, (count_input_dims_1*sizeof(int64_t)));
+
+    status = vxUnmapTensorPatch((vx_tensor)parameters[1], map_id);
+    if(status) 
+    {
+        std::cerr << "ERROR: vxUnmapTensorPatch() failed for input#5 (" << status << ")" << std::endl;
+        return -1;
+    }
+
+    vx_int32 axis;
+    ERROR_CHECK_STATUS(vxCopyScalar((vx_scalar)parameters[2], &axis, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+
+    //opposite of onnx. Maybe have to remove for model. Works with GDF testing
+    axis = 3 - axis;
+    
+    //accessing the tensor from last dimension
+    if(axis > 3)
+    {
+        int count = 0, dims_idx = 0;
+        //get real dimensions. Ignoring the appended 1s
+        while(dims_idx <= 3 && input_dims_0[dims_idx] == 1)
+        {
+            count++;
+            dims_idx++;
+        }
+
+        count = 4 - count;
+        axis = axis - count;
+    }
+
+    vx_int32 largest;
+    ERROR_CHECK_STATUS(vxCopyScalar((vx_scalar)parameters[3], &largest, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+
+    vx_int32 sorted;
+    ERROR_CHECK_STATUS(vxCopyScalar((vx_scalar)parameters[4], &sorted, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+
+    //output vectors
+    std::vector<float> values;
+    std::vector<int64_t> indices;
+
+    //vector to sort indices
+    std::vector<size_t> idx(input_dims_0[axis]);
+    
+    for(int i = 0; i < count_input_dims_0; i += input_dims_0[axis])
+    {
+        std::iota(idx.begin(), idx.end(), 0); 
+        if (largest)
+        {   if (sorted)

```

Comment:
move if (sorted) to next line. Also do we really need 2 types of sort. std::stable_sort is less efficient?

Reply:
I retained only std::sort. Doesn't require it by onnx.